js/ui/grid_core/ui.grid_core.column_chooser.js

```diff
@@ -318,23 +294,67 @@ const columnChooserMembers = {
 
     _prepareSelectModeConfig: function() {
         const that = this;
-        const selectionChangedHandler = function(e) {
-            const visibleColumns = that._columnsController.getVisibleColumns().filter(function(item) { return !item.command; });
-            const isLastColumnUnselected = visibleColumns.length === 1 && !e.itemData.selected;
+        const selectionOptions = this.option('columnChooser.selection') || {};
 
-            if(isLastColumnUnselected) {
-                e.component.selectItem(e.itemElement);
-            } else {
-                setTimeout(function() {
-                    that._columnsController.columnOption(e.itemData.id, 'visible', e.itemData.selected);
+        const flatNodes = (nodes, result = []) => {
+            nodes.forEach(node => {
+                result.push(node);
+
+                if(node.children.length) {
+                    flatNodes(node.children, result);
+                }
+            });
+
+            return result;
+        };
```

Comment:
Let's change this function:
- Incapsulate default ```result``` argument, to prevent incorrect usage of this function in the future :)
- Let's use ```reduce```
- Let's rename this function to ```getFlatNodes```

```suggestion
    const getFlatNodes = (nodes) => {
      const addNodesToArray = (nodes, flatNodesArray) => {
        return nodes.reduce((result, node) => {
            result.push(node);

            if(node.children.length) {
                addNodesToArray(node.children, result);
            }

            return result;
        }, flatNodesArray);
    };

    return addNodesToArray(nodes, []);
};
```